### ISSUE TITLE
Unable to add a custom Decoder using ConfigLoader.Builder

### ISSUE THREAD

**Issue by smcbn** [sksamuel/hoplite#157]
Hi,

I am attempting to add a custom decoder like so:
```
val configLoader = ConfigLoader.Builder().addDecoder(CustomDecoder()).build()
val config = configLoader.loadConfigOrThrow(configFile)
```
However the `CustomDecoder` never gets added to the `decoderRegistry` that is used to construct the `ConfigLoader` in `Builder.build()`:

The area of code is [here](https://github.com/sksamuel/hoplite/blob/master/hoplite-core/src/main/kotlin/com/sksamuel/hoplite/ConfigLoader.kt#L170). The return value from `decoderRegistry.register(decoder)` is ignored and `decoderRegistry` never is updated to contain the additional decoder.

I believe the code should be:

```
      var decoderRegistry = defaultDecoderRegistry(this.classLoader)
      this.decoderStaging.forEach { decoder: Decoder<*> ->
        decoderRegistry = decoderRegistry.register(decoder)
      }
```

A workaround I have is to instantiate the ConfigLoader myself:
```
        val classLoader = Thread.currentThread().contextClassLoader
        val configLoader = ConfigLoader(
                decoderRegistry = defaultDecoderRegistry(classLoader).register(CustomDecoder()),
                propertySources = defaultPropertySources(defaultParserRegistry(classLoader)),
                parserRegistry = defaultParserRegistry(classLoader),
                preprocessors = defaultPreprocessors(),
                paramMappers = defaultParamMappers()
        )
```
The above code results in my `CustomDecoder` being used.

**Comment by AvlWx2014** [comment]
Hi @smcbn I think your suggestion will fix this. I will incorporate it and submit a PR for it. Good catch! It looks like this is also going to be an issue [here](https://github.com/sksamuel/hoplite/blob/master/hoplite-core/src/main/kotlin/com/sksamuel/hoplite/ConfigLoader.kt#L177).

I'm thinking there's a way we can do this without creating a new instance of `DecoderRegistry` on each iteration and reassigning `decoderRegistry` repeatedly. 

I wonder what @sksamuel thinks about exposing the list of decoders from `DecoderRegistry` rather than keeping them private like this
```kotlin
interface DecoderRegistry {
  val decoders: List<Decoder<*>>
  // rest of interface
}

class DefaultDecoderRegistry(override val decoders: List<Decoder<*>>) : DecoderRegistry {
  // class body
}
```

Then we can do something like this instead:
```kotlin
val defaultDecoders = defaultDecoderRegistry(this.classLoader)
val decoderRegistry = DefaultDecoderRegistry(defaultDecoders.decoders + this.decoderStaging)
```

**Comment by sksamuel** [comment]
I don't think there's a great harm in doing things in the functional way and copying, since config is loaded once.
So in order to fix this bug quickly, I've changed the offending line to this:

```
val decoderRegistry = this.decoderStaging.fold(defaultDecoderRegistry(this.classLoader)) { registry, decoder ->
  registry.register(decoder)
}
```

And released 1.3.4.

@AvlWx2014 if you want to make a PR to expose the decoder registry interface, feel free.

**Comment by AvlWx2014** [comment]
This is a great approach! `fold` is one among a group of builtin functions in Kotlin I don't use often that slip my mind. I don't think anything else is necessary.
